### PR TITLE
new components for easier question and result loading

### PR DESCRIPTION
**Pull request by kdoh**
### The tl;dr; 
```jsx
const MyGoodNewFeature = ({
  // these probably come from the router, but they could be set by you too
  questionId,
  questionHash, 
}) =>
<QuestionAndResultLoader questionId={questionId} questionHash={questionHash}>
{ ({ question, result, cancel, reload }) =>
  <div>{ question.displayName() }</div>
  { result && ( <Visualization rawData={[{ card. question.card(), data: result.data }]} />)
  <a onClick={() => cancel()}>Cancel this query please</a>
  <a onClick={() => reload()}>Freshen this up for me</a>
}
</QuestionAndResultLoader>
```

[A more real example of this in use can be seen here.
](https://github.com/metabase/metabase/blob/entity-pages/frontend/src/metabase/components/EntityPage.jsx)

### Motivation
A few important things need to happen when you want to pull down a question. First you need to get the question either from the API or the URL (if it's ad-hoc), then you need to make sure you have metadata about the source table, or db for said loaded, and then to manipulate it nicely with metabase-lib you need to instantiate a new Question object. If you want to work with the result, you also need to call a method on that question object and then if you want to reload or cancel that... 

These steps weren't really fully documented or easily discoverable without spending a fair bit of time or having familiarity with the metabase codebase.

@tlrobinson and I decided to clean this up a bit by creating some convenience components to help with Saved Question loading, Ad-Hoc question loading, result loading, and then for convenience, handling all those situations in one simple component. Props to Tom for the clever way this is split out.

Each component can be used on its own for cases where you might only care about information about a Saved Question or together in cases where you're doing something more custom, but the power for prototyping or new features comes from `<QuestionAndResultLoader>` which combines everything together to give you your question, the result, and basic controls for canceling or reloading the query.

### Future thoughts
Eventually I'd love it if this also handed loading state as well, right now you need to add something akin to: 
```jsx
if(!question) {
  return <div>Loading...</div>
}
```
to make sure that you aren't trying to access information that isn't there yet.

### CODE REVIEW

File: frontend/src/metabase/containers/AdHocQuestionLoader.jsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * @example
+ *
+ * The raw un-connected component is also exported so we can unit test it
+ * without the redux store.
+ */
+
+type Props = {
+  children: Function,
+  loadMetadataForCard: Function,
```
Comment (tlrobinson) [inline frontend/src/metabase/containers/AdHocQuestionLoader.jsx]: These are ok but if you want to be more precise you can do `(card: Card) => Promise<void>`

File: frontend/src/metabase/containers/AdHocQuestionLoader.jsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    // pass the decoded card to load any necessary metadata
+    // (tables, source db, segments, etc) into
+    // the redux store, the resulting metadata will be avaliable as metadata on the
+    // component props once it's avaliable
+    await this.props.loadMetadataForCard(card);
+
+    // instantiate a new question object using the metadata and saved question
+    // so we can use metabase-lib methods to retrieve information and modify
+    // the question
+    const question = new Question(this.props.metadata, card);
```
Comment (tlrobinson) [inline frontend/src/metabase/containers/AdHocQuestionLoader.jsx]: One thing I'm realizing now is that if the `metadata` is later updated after `question` is added to the state then it won't update with the new metadata. This probably won't happen often but we should handle it.

There are a few ways

1. Store both `question` and `card` in `state`, add the following to `componentWillReceiveProps`:
```
if (nextProps.metadata !== this.props.metadata && this.state.card) {
  this.setState({ question: this.state.card && new Question(nextProps.metadata, this.state.card) })
}
```

2. Only store `card` in `state`, create a reselect selector for `question` in the constructor:
```
this._getQuestion = createSelector(
  [(s, p) => s.card, (s, p) => p.metadata],
  (card, metadata) => card && new Question(metadata, card)
)
```
then in `render` do
```
return this.props.children(this._getQuestion(this.state, this.props));
```

3. Implement `shouldComponentUpdate` (or possibly just use `React.PureComponent` but see this caveat https://reactjs.org/docs/render-props.html#be-careful-when-using-render-props-with-reactpurecomponent) to check if `card` or `metadata` has changed:
```
return nextProps.metadata !== this.props.metadata || nextState.card !== this.state.card
```
and in `render` just do:
```
return this.props.children(this.state.card && new Question(this.props.metadata, this.state.card));
```

File: frontend/src/metabase/containers/QuestionResultLoader.jsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  componentWillMount() {
+    this._loadResult(this.props.question);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // if the question is different, we need to do a fresh load
+    // TODO - is this the best way to check for new questions? any performance
+    // implications?
+    if (nextProps.question !== this.props.question) {
```
Comment (tlrobinson) [inline frontend/src/metabase/containers/QuestionResultLoader.jsx]: It might be best to do a "deeper" comparison in case a calling component creates a new instead of `Question` but it hasn't actually changed.

A simple check could be:

```
if ((nextProps.question && nextProps.question.getUrl()) !== (this.props.question && this.props.question.getUrl))
```

I *think* that would work.

We also might want to eventually add a `noReload` prop or something to prevent automatic reloading, but if we think the default should be to reload then we can leave it out until we need it.

File: frontend/src/metabase/containers/SavedQuestionLoader.jsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    // (tables, source db, segments, etc) into
+    // the redux store, the resulting metadata will be avaliable as metadata on the
+    // component props once it's avaliable
+    await this.props.loadMetadataForCard(card);
+
+    // instantiate a new question object using the metadata and saved question
+    // so we can use metabase-lib methods to retrieve information and modify
+    // the question
+    //
+    const question = new Question(this.props.metadata, card);
```
Comment (tlrobinson) [inline frontend/src/metabase/containers/SavedQuestionLoader.jsx]: Same issue I mentioned in `AdHocQuestionLoader` regarding updating `question` when `metadata` changes.

File: frontend/test/containers/AdHocQuestionLoader.unit.spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    shallow(
+      <AdHocQuestionLoader questionHash={questionHash}>
+        {() => <div />}
+      </AdHocQuestionLoader>,
+    );
+
+    expect(loadSpy).toHaveBeenCalledWith(questionHash);
+  });
+
+  it("should load a new question if the question ID changes", () => {
```
Comment (tlrobinson) [inline frontend/test/containers/AdHocQuestionLoader.unit.spec.js]: s/question ID/question hash/

File: frontend/test/containers/AdHocQuestionLoader.unit.spec.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  it("should load a question given a questionHash", () => {
+    const questionHash = "#abc123";
+
+    const loadSpy = jest.spyOn(AdHocQuestionLoader.prototype, "_loadQuestion");
+
+    shallow(
+      <AdHocQuestionLoader questionHash={questionHash}>
+        {() => <div />}
```
Comment (tlrobinson) [inline frontend/test/containers/AdHocQuestionLoader.unit.spec.js]: It would be good to check to see if the render prop/children method is called with the expected `question` object (perhaps asserting the output of `question.getUrl()` or the  `question.card()`, something like

```javascript
const mockRenderProp = jest.fn().mockReturnValue(<div />);
shallow(
  <AdHocQuestionLoader questionHash={questionHash}> children={mockRenderProp} />
);
const calls = mockRenderProp.mock.calls;
expect(calls[calls.length -  1][0].card()).toEqual(EXPECTED_CARD_OBJECT);
```

**Review by tlrobinson — CHANGES_REQUESTED** [review]
Looks great overall. Just a few things mentioned in the comments.

File: frontend/src/metabase/containers/AdHocQuestionLoader.jsx (hunk 2 above)
Comment (kdoh) [reply frontend/src/metabase/containers/AdHocQuestionLoader.jsx]: Do you have a preference for which approach we should take? It seems like 1 might be the easiest to grok.
